rfcs/0009-abstraction-layer-for-mobile-error-tracking.md

```diff
@@ -0,0 +1,191 @@
+
+Another important motivation is to add the ability to toggle the error reporting whenever we want to. Unfortunately, some solutions don't have this feature built-in, so we will create a way to handle this inside the abstraction layer.
+
+# Detailed design
+
+Example for Sentry:
+
+```js
+import * as Sentry from "@sentry/react-native";
+
+const captureError = Sentry.captureException;
+
+// level should be: fatal, error, warning, info, or debug.
+const addBreadcrumb = ({ scope, message, level = "info" }) =>
+  Sentry.addBreadcrumb({
+    category: scope,
+    message,
+    level,
+  });
+
+const setUserId = (level) =>
+  Sentry.configureScope((scope) => scope.setLevel(level));
+
+// You can set the severity of an event to one of five values: fatal, error, warning, info, and debug. error is the default, fatal is the most severe, and debug is the least severe.
+
+const setLevel = (id) =>
+  Sentry.configureScope((scope) => scope.setUser({ id }));
+
+const enable = (enabled) => {
+  isTrackingEnabled = enabled;
+};
+
+const createErrorTracking = ({ enabled = false } = {}) => {
+  let isTrackingEnabled = enabled;
+
+  const createTrackingFn = (fn) => (...args) => {
+    if (!isTrackingEnabled) {
+      return Promise.resolve();
+    }
+
+    return fn(...args);
+  };
+
+  return {
+    captureError: createTrackingFn(captureError),
+    addBreadcrumb: createTrackingFn(addBreadcrumb),
+    setUserId: createTrackingFn(setUserId),
+    setLevel: createTrackingFn(setLevel),
+    enable,
+  };
+};
+
+export default createErrorTracking;
+```
+
+Example for Firebase Crashlytics:
+
+```js
+import crashlytics from "@react-native-firebase/crashlytics";
+
+const captureError = crashlytics().recordError;
+
+const setUserId = crashlytics().setUserId;
+
+const setLevel = (level) => crashlytics().setAttributes({ level });
+
+// level should be: fatal, error, warning, info, or debug.
+const addBreadcrumbs = ({ scope, message, level = "info" }) =>
+  crashlytics().setAttributes({ category: scope, message, level });
+
+const enable = (enabled) => {
+  isTrackingEnabled = enabled;
+};
+
+const createErrorTracking = ({ enabled = false } = {}) => {
+  let isTrackingEnabled = enabled;
+
+  const createTrackingFn = (fn) => (...args) => {
+    if (!isTrackingEnabled) {
+      return Promise.resolve();
+    }
+
+    return fn(...args);
+  };
+
+  return {
+    captureError: createTrackingFn(captureError),
+    addBreadcrumbs: createTrackingFn(addBreadcrumbs),
+    setLevel: createTrackingFn(setLevel),
+    setUserId: createTrackingFn(setUserId),
+    enable,
+  };
+};
+
+export default createErrorTracking;
+```
+
+## Usage examples
+
+```js
+import createErrorTracking from "@moxy/react-native-sentry";
+
+const errorTracking = createErrorTracking({ isEnable: true });
+
+try {
+  aFunctionThatMightFail();
+} catch (err) {
+  errorTracking.captureError(err);
+}
+```
+
+## Regarding the initialization
```

Comment:
I don't see why we have to delegate initialization to the developer. From what I've learned, both Sentry and BugSnag use a plain object with several options, which is a very standard approach to handle configuration. Thus, I'd say we can simply accept such options in the factory function and forward them to the specific initialization function.

Reply:
I thought that Bugsnag was basically out of the game, since we realized that Crashlytics should suite small things, while Sentry would do for the bigger ones because it has a better pricing in comparison to Bugsnag. That said, Sentry and Crashlytics are completely different, as I've exposed here.

Reply:
For the time being we're not planning to use BugSnag, but we never know... Regardless, with or without it, the same rationale applies. Crashlytics does not require any initialization on JS side? Great! Nothing has to be done. If it ever does, we can add it later. Sentry takes an object with a few options which I believe a reasonable approach we can adopt for whatever solution we may end up using.
If we move the initialization logic up to the caller, we have to import the SDK of the solution in the app in addition to importing in the implementation modules themselves, so we end up leaking implementation details because the app has to know which SDK to import.
If we effectively go down the road of performing initialization in the app, it would probably be better to declare the SDK as a peer dependency in the implementation modules, install it manually in the app and pass it as an argument to the error tracking factory function.

Reply:
We decided to go for peer dependencies to have more control over the initialization process, which becomes a responsibility of the app, and the version of the SDK. The instance of the SDK is passed as an argument to the implementation module.